group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```
@@ -14,17 +14,903 @@
+                }
+            }
+        }
+        unassignedPartitions.removeAll(fixedPartitions);
+
+        // Narrow down the reassignment scope to only those members that are subject to reassignment.
+        for (String member : members.keySet()) {
+            if (!canMemberParticipateInReassignment(member)) {
+                sortedMembersByAssignmentSize.remove(member);
+            }
+        }
+
+        // If all the partitions are fixed i.e. unassigned partitions is empty there is no point of re-balancing.
+        if (!unassignedPartitions.isEmpty()) performReassignments();
+    }
+
+    private void performReassignments() {
+        boolean modified;
+        boolean reassignmentOccurred;
+        // Repeat reassignment until no partition can be moved to improve the balance.
+        do {
+            // Before re-starting the round of reassignments check if the assignment is already balanced.
+            if (isBalanced()) break;
+
+            modified = false;
+            reassignmentOccurred = false;
+            // Reassign all reassignable partitions sorted in descending order
+            // by totalPartitions/number of subscribed members,
+            // until the full list is processed or a balance is achieved.
+            List<TopicIdPartition> reassignablePartitions = sortTopicIdPartitions(unassignedPartitions);
+
+            for (TopicIdPartition reassignablePartition : reassignablePartitions) {
+                // Only check if there is any change in balance if any moves were made.
+                if (reassignmentOccurred && isBalanced()) {
+                    return;
+                }
+                reassignmentOccurred = false;
+
+                // The topicIdPartition must have at least two consumers.
+                if (membersPerTopic.get(reassignablePartition.topicId()).size() <= 1)
+                    LOG.error("Expected more than one potential member for topicIdPartition '{}'", reassignablePartition);
```

nit: two members

should we throw an exception here? we don't expect any partition to have less than 2 members since we already removed all of them before calling performReassignments() right?

---

yes that's right, changed it to an exception